### PR TITLE
Add a use:stable command, inverse of use:dev

### PR DESCRIPTION
**Pull request by csirmazbendeguz**
I used https://updates.drupal.org to parse release information.
The latest stable release version is decided by comparing the release versions to the current composer constraint.

### CODE REVIEW

File: Commands/ReleaseHistoryClient.php [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+   *   The name of the Drupal project (e.g. 'lightning_core').
+   * @param string $api_version
+   *   The API compatibility version (e.g. '8.x').
+   * @param string $range
+   *   The constraint range (e.g. '3.x-dev').
+   *
+   * @return string
+   *   The latest stable release (e.g. '3.2'), or empty string if not found.
+   */
+  public function getLatestStableRelease($name, $api_version, $range) {
```
Comment (phenaproxima) [inline Commands/ReleaseHistoryClient.php]: We're only supporting Drupal 8, so I don't think we need the $api_version parameter here. We can just hard-code 8.x. If you wanna get weird, you can also use the \Drupal::CORE_COMPATIBILITY constant, which is hard-coded as 8.x.

File: Commands/ReleaseHistoryClient.php [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    $release_history = $this->getReleaseHistory($name, $api_version);
+
+    $prefix = "$api_version-";
+    $prefix_length = strlen($prefix);
+
+    // Remove the '.x-dev' suffix from range if present.
+    $suffix = '.x-dev';
+    $suffix_position = -strlen($suffix);
+
+    if (substr($range, $suffix_position) === $suffix) {
```
Comment (phenaproxima) [inline Commands/ReleaseHistoryClient.php]: Not a big deal, but couldn't we do this with a regex?

File: Commands/ReleaseHistoryClient.php [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+      $range = substr($range, 0, $suffix_position);
+    }
+
+    // Releases are ordered from newest to oldest.
+    foreach ($release_history->releases->release as $release) {
+      $release_version = (string) $release->version;
+
+      // Remove the API version prefix from the release version if exists.
+      // For example, if the release version is '8.x-3.2', it will be '3.2'.
+      if (strncmp($release_version, $prefix, $prefix_length) === 0) {
```
Comment (phenaproxima) [inline Commands/ReleaseHistoryClient.php]: Same here; feels like a regex would suffice?

File: Commands/ReleaseHistoryClient.php [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+   *
+   * @param string $name
+   *   The name of the Drupal project (e.g. 'lightning_core').
+   * @param string $api_version
+   *   The API compatibility version (e.g. '8.x').
+   *
+   * @return \SimpleXMLElement
+   *   The parsed XML response.
+   */
+  private function getReleaseHistory($name, $api_version) {
```
Comment (phenaproxima) [inline Commands/ReleaseHistoryClient.php]: No need for $api_version, let's just hard-code 8.x.

File: Commands/RoboFile.php [hunk 5]
```diff
@@ -642,4 +643,47 @@ public function useDev ()
...
+        foreach ($composer['require'] as $package => $constraint)
+        {
+            if ($package === 'drupal/core')
+            {
+                $callback = function ($range) use ($client) {
+                    $release = $client->getLatestStableRelease('drupal', '8.x', $range);
+
+                    return $release ? "~$release" : $range;
+                };
+                $version = (new ComposerConstraint($constraint))->mapRanges($callback);
```
Comment (phenaproxima) [inline Commands/RoboFile.php]: These two lines ($version = .... and $task->dependency()) are repeated. Can we do them after the elseif?

File: Commands/RoboFile.php [hunk 6]
```diff
@@ -642,4 +643,47 @@ public function useDev ()
...
+                $task->dependency($package, $version);
+            }
+            elseif (strpos($package, 'drupal/lightning_') === 0)
+            {
+                $name = strstr($package, 'lightning_');
```
Comment (phenaproxima) [inline Commands/RoboFile.php:676]: Never saw this use of strstr before. Nice one :)

File: tests/src/Unit/ReleaseHistoryClientTest.php [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+   *   The constraint range.
+   * @param string $release_history
+   *   The mock response body.
+   * @param string $expected
+   *   The expected result.
+   *
+   * @dataProvider getLatestStableReleaseProvider
+   */
+  public function testGetLatestStableRelease($name, $api_version, $range, $release_history, $expected) {
+    $client = $this->getMockBuilder(Client::class)
```
Comment (phenaproxima) [inline tests/src/Unit/ReleaseHistoryClientTest.php]: Can we use $this->prophesize() to create mocks, rather than $this->getMockBuilder()?

**Review by phenaproxima — CHANGES_REQUESTED** [review]
I think this looks excellent. I'd like to manually test it a bit, but honestly I can't wait to merge it. This is the major blocker to automated releases, so I'm very excited!!

File: Commands/ReleaseHistoryClient.php (hunk 2 above)
Comment (csirmazbendeguz) [reply Commands/ReleaseHistoryClient.php]: You are right, it's looks much cleaner with regex.

File: Commands/ReleaseHistoryClient.php (hunk 1 above)
Comment (csirmazbendeguz) [reply Commands/ReleaseHistoryClient.php]: Yeah, it was hardcoded anyway, just on the command level.

File: Commands/RoboFile.php (hunk 6 above)
Comment (csirmazbendeguz) [reply Commands/RoboFile.php:676]: Thanks! It's not a commonly used function in PHP.

**Review by phenaproxima — APPROVED** [review]

**Comment by phenaproxima** [comment]
Will manually test this a bit, then likely merge. Thanks, @csirmazbendeguz!

**Comment by phenaproxima** [comment]
Tested manually a bit and I gotta tell you, this worked great. Wow. This is going to make release days that much less painful. Thank you!

**Pull request merged**